### PR TITLE
Allow group changes in composed templates

### LINKED ISSUE #5473: v1.14/v1.15 - cannot change the kind or group of a composed resource
### What happened?
We're working on moving our resources from the community AWS provider (https://github.com/crossplane-contrib/provider-aws) to the official Upbound AWS provider (https://github.com/upbound/provider-aws). 

Here's a brief overview of the steps involved:
1) Add `crossplane.io/paused` annotation and `deletionPolicy: Orphan` on all community managed resources
2) Shift all managed resources to the official provider.
3) In the Git repository, update the composition but keep the `order` and `spec.resources[].name` unchanged.
4) Apply the updated composition.
5) The composition will detect all new resources and update `resourceRefs[].apiVersion` and `resourceRefs[].Kind` in the **XR**, while the resource names remain the same.
6) Remove all community managed resources from the community AWS provider.

This process worked with Crossplane version `v1.13.x` but encountered issues starting with versions `v1.14.x`/`v1.15.x`. The error message encountered is:

```
cannot compose resources: cannot parse base template of composed resource 'iam': cannot change the kind or group of a composed resource from iam.aws.crossplane.io/v1beta1, Kind=Role to iam.aws.upbound.io/v1beta1, Kind=Role (possible composed resource template mismatch)
```

The change in behavior with Crossplane versions `v1.14.x` and `v1.15.x` raises questions, especially since migrating to the official provider is a common use case.

change of the behavior implemented with: https://github.com/crossplane/crossplane/pull/4500
recent change for version updates: https://github.com/crossplane/crossplane/pull/5369
discussion between @phisco and @turkenh  https://github.com/crossplane/crossplane/pull/5369#discussion_r1486019598

### How can we reproduce it?
https://github.com/haarchri/xp-issue-5473

### What environment did it happen in?
Crossplane version: v1.14.6-up.1, v1.15.0-up.1

### PR DESCRIPTION
**Pull request by turkenh**
### Description of your changes

As @phisco successfully foresaw in [this comment](https://github.com/crossplane/crossplane/pull/5369#discussion_r1486019598) 🙌 , checking the group backfired during migrations. This PR reverts the check introduced in [this PR ](https://github.com/crossplane/crossplane/pull/4500) and only leaves [checking the kind](https://github.com/crossplane/crossplane/blob/cb232d32a595876a98ab3a7f5125e3fa26e2f2ac/internal/controller/apiextensions/composite/composition_pt.go#L512) as in prior to v1.13.

Fixes #5473

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [ ] ~Added or updated e2e tests.~
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [x] Added `backport release-x.y` labels to auto-backport this PR.

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

**Review by phisco — APPROVED** [review]
In Italy we say that also broken clocks tell the right time twice a day 🇮🇹🕰️😂

File: internal/controller/apiextensions/composite/composition_render.go
Comment (negz) [inline internal/controller/apiextensions/composite/composition_render.go:76]: Changing the API version (e.g. v1beta1 -> v1) of a composed resource can be done in place. Crossplane will just start hitting the new v1 API endpoint to update the state.

You can't update the API group in place, as far as I know. So if you change the API group here Crossplane will:

* Create a new composed resource in the new API group
* Treat the old composed resource as no longer desired, and delete it

This enables migrating to a different API group, but the migration will be destructive. Are we comfortable with that?

Comment (negz) [reply internal/controller/apiextensions/composite/composition_render.go:76]: By enabling (group) migrations we also weaken a safety check.

Consider especially 'legacy' compositions where the composed templates don't have a `name`. Crossplane must then use the order of the `resources` array to figure out which template corresponds to which existing composed resource. If folks reorder that list (or remove an element from the middle) unexpected things would happen (e.g. composed resources would be deleted and recreated).

To be fair, I can see a case that just checking for `Kind` is good enough. It's unlikely that you accidentally switch two composed resources that are different but have the same `Kind`.

On the other hand, maybe it would be more explicit when doing a migration to a completely different kind of composed resource to change its composed resource name. This might make it more obvious that the existing composed resource will be deleted and replaced with a new one, not updated in place.

**Review by negz — APPROVED** [review]
Approving to unblock, but please consider my comment thread before merging.

File: internal/controller/apiextensions/composite/composition_render.go
Comment (haarchri) [reply internal/controller/apiextensions/composite/composition_render.go:76]: We've running the migration process across several environments for a few months, and this process halted with Crossplane v1.14/v1.15 - We want the simplified method back to update composition to official provider resources without manually updating resourceRefs ( 120 resources * 25 Claims/XRs) or other migration steps. We have all official managed resources pre-created in an earlier step with same metadata.name, labels and annotations - the only step missing is now that the XR starts managing these pre-created official managed resources - which was working with v1.13. While you acknowledge the issues with real legacy compositions lacking `spec.resources[].name` - i understand this - but in real most of the environments using the `name` - and in this case, we're using the name to adopt the official managed resource

Comment (turkenh) [reply internal/controller/apiextensions/composite/composition_render.go:76]: I agree with the points @negz is making about weakening a safety check, on the other hand, I feel like that they are not that crucial in practice. My main argument for that is that we are running without such a check, up to the Crossplane version 1.13, and there are no known incidents related to that (AFAIU). 
So, I am leaning towards moving forward with this PR to facilitate the migration scenarios that @haarchri is pointing out.

**Pull request merged**
